runtime/jailer.go

```diff
@@ -0,0 +1,476 @@
+				int(j.Config.UID),
+				int(j.Config.GID),
+			); err != nil {
+				return errors.Wrapf(err, "failed to chown -r %v", rootPath)
+			}
+
+			j.logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.ContentsPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// exposeFileToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail.
+func (j runcJailer) ExposeMountToJail(srcMountPath string) error {
+	uid := j.Config.UID
+	gid := j.Config.GID
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcMountPath, &stat); err != nil {
+		return err
+	}
+
+	path := j.RootPath()
+	dst := ""
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK {
+		path = filepath.Join(path, "dev")
+		dst = filepath.Join(path, filepath.Base(srcMountPath))
+		if err := exposeBlockDeviceToJail(dst, int(stat.Rdev)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, int(uid), int(gid)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, int(rdev)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	if info, err := os.Stat(dst); err == nil {
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	} else if os.IsNotExist(err) {
+		// this occurs when a full path is provided including the file name
+	} else if err != nil {
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY, mode)
```

Comment:
We should probably include `O_EXCL` here to ensure we are the ones creating the file. Otherwise if there is a pre-existing file there, I think we will end up writing from the start of the file, but if the original file is larger than what we are writing, some of the data from the original file will be left at the end (which is either a bug or a security vulnerability).